fairscale/optim/oss.py

```diff
@@ -153,12 +153,22 @@ def step(self, closure: Optional[Callable[[], float]] = None, **kwargs: Any) ->
 
         # Sync all the states. Broadcast requests are issued async, we check completeness before moving on
         requests = []
+        requires_grad = []
         for rank, param_groups in enumerate(self.partition_parameters()):
             for param_group in param_groups:
                 for param in param_group["params"]:
+                    # NOTE: Broadcast is in-place and not differentiable
+                    # Gloo will rightly assert on this operation for any tensor that requires grad.
+                    # We save and restore the grad requirement state to work around that, in our case
+                    # the grad is only useful on the source rank.
+                    requires_grad.append((param, param.requires_grad))
+                    param.requires_grad = False
                     requests.append(dist.broadcast(tensor=param, src=rank, group=self.group, async_op=True))
```

Comment:
it seems that you can chain the assignment using future.then() function with a lambda function?

Reply:
ah I didn't know that it was supported with the futures that pytorch returns ? I'll have a look

Reply:
also here: https://pytorch.org/docs/stable/futures.html#torch.futures.collect_all

wait_all() can be used in place of a loop.

Reply:
yes for wait_all, I guess that it's a loop under the hood but cleaner indeed ! I"m not sure that the then() formulation proposed will work though, lambdas are stateless in python so you typically cannot change a member value

Reply:
I thought you can pass in to `then()` a closure which contains the reference to the `param` variable, then you can freely change its memory the original value. It could avoid the extra list entirely. But I could be wrong.

Reply:
not possible unfortunately, I just tried, the futures linked above are from the RPC framework, not pytorch.distributed

Reply:
the only methods supported [here](https://pytorch.org/docs/stable/distributed.html#synchronous-and-asynchronous-collective-operations) are .wait() and is_completed()